Reduce number of coin flips in RecordFunction

ilia-cher opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#40758 [rfc] Reduce number of coin flips in RecordFunction**

Summary:
Currently we flip a coin for each sampled callback each time
we run RecordFunction, this PR is an attempt to skip most of the coin
flips (for the low-probability observers) and keep the distribution
close to the original one

Test Plan:
CI and record_function_benchmark
```
(python_venv) iliacher@devgpu151:~/local/pytorch  (reduce_coin_flops)$ ./build/bin/record_function_benchmark
Warmup time: 30108 us.
Time per iteration (1x1): 1496.78 us.
Time per iteration (16x16): 2142.46 us.
Pure RecordFunction runtime of 10000000 iterations 687929 us, number of callback invocations: 978
(python_venv) iliacher@devgpu151:~/local/pytorch  (reduce_coin_flops)$ ./build/bin/record_function_benchmark
Warmup time: 19051 us.
Time per iteration (1x1): 1581.89 us.
Time per iteration (16x16): 2195.67 us.
Pure RecordFunction runtime of 10000000 iterations 682402 us, number of callback invocations: 1023
(python_venv) iliacher@devgpu151:~/local/pytorch  (reduce_coin_flops)$ ./build/bin/record_function_benchmark
Warmup time: 18715 us.
Time per iteration (1x1): 1566.11 us.
Time per iteration (16x16): 2131.17 us.
Pure RecordFunction runtime of 10000000 iterations 693571 us, number of callback invocations: 963
(python_venv) iliacher@devgpu151:~/local/pytorch  (reduce_coin_flops)$
(python_venv) iliacher@devgpu151:~/local/pytorch  (reduce_coin_flops)$ ./build/bin/record_function_benchmark
Warmup time: 21580 us.
Time per iteration (1x1): 1912.39 us.
Time per iteration (16x16): 2049.49 us.
Pure RecordFunction runtime of 10000000 iterations 651399 us, number of callback invocations: 999, expected number: ~1000 invocations
(python_venv) iliacher@devgpu151:~/local/pytorch  (reduce_coin_flops)$ ./build/bin/record_function_benchmark
Warmup time: 18947 us.
Time per iteration (1x1): 1591.05 us.
Time per iteration (16x16): 2199.92 us.
Pure RecordFunction runtime of 10000000 iterations 665978 us, number of callback invocations: 1014, expected number: ~1000 invocations
(python_venv) iliacher@devgpu151:~/local/pytorch  (reduce_coin_flops)$ ./build/bin/record_function_benchmark
Warmup time: 17512 us.
Time per iteration (1x1): 1455.78 us.
Time per iteration (16x16): 2055.89 us.
Pure RecordFunction runtime of 10000000 iterations 676929 us, number of callback invocations: 988, expected number: ~1000 invocations

before the PR coin flipping change:

(python_venv) iliacher@devgpu151:~/local/pytorch  (reduce_coin_flops)$ ./build/bin/record_function_benchmark
Warmup time: 18814 us.
Time per iteration (1x1): 1536.2 us.
Time per iteration (16x16): 1985.82 us.
Pure RecordFunction runtime of 10000000 iterations 944959 us, number of callback invocations: 1015
(python_venv) iliacher@devgpu151:~/local/pytorch  (reduce_coin_flops)$ ./build/bin/record_function_benchmark
Warmup time: 18278 us.
Time per iteration (1x1): 1526.32 us.
Time per iteration (16x16): 2093.77 us.
Pure RecordFunction runtime of 10000000 iterations 985307 us, number of callback invocations: 1013
(python_venv) iliacher@devgpu151:~/local/pytorch  (reduce_coin_flops)$ ./build/bin/record_function_benchmark
Warmup time: 18545 us.
Time per iteration (1x1): 1524.65 us.
Time per iteration (16x16): 2080 us.
Pure RecordFunction runtime of 10000000 iterations 952835 us, number of callback invocations: 1048
```

Differential Revision: [D22320879](https://our.internmc.facebook.com/intern/diff/D22320879)

Commit:
[rfc] Reduce number of coin flips in RecordFunction

Summary:
Currently we flip a coin for each sampled callback each time
we run RecordFunction, this PR is an attempt to skip most of the coin
flips (for the low-probability observers) and keep that the distribution
close to the original one

Test Plan:
CI and record_function_benchmark

[ghstack-poisoned]

ngimel commented on aten/src/ATen/record_function.cpp:
```diff
@@ -222,6 +233,38 @@ double RecordFunctionCallback::sample_zero_one() {
+    // sampled from the geometric distribution
+    if (sampling_prob_ < kLowProb) {
+      if (tries_left_ == 0) {
+        tries_left_ = sample_geometric();
```
you just need 
```
if (tries_left_ == 0){
 tries_left_ = sample_exponential()/sampling_prob_ + .5;
 return true;
} else {...
```
here. Wait time until the next event in poisson process is exponential distribution http://homepage.divms.uiowa.edu/~rdecook/stat2020/notes/ch4_pt3.pdf, and `+.5` is so that you round to nearest when you convert to int instead of round down (you could also use std::round).

ilia-cher replied:
the issue here is that different callbacks have different sampling_prob_ so I guess we can't use a simple thread local tries_left_ in this case

dzhulgakov replied:
would it be hard to just have per-callback counter? I'm always worried about mixing different distributions.

If we indeed have only a single callback usually, I'm fine landing it as is and following up

dzhulgakov reviewed (commented):
Can we add a simple test/benchmark that runs record function many many times with several callbacks and different probabilities and validates the the number of times they are triggered matches expection on average?

ilia-cher commented:
we have record_function_benchmark that does that (runs the loop with observers setup), I'll add the result in the test section

I think the problem with per callback counter is that we would need to sync on its decrement for the global callbacks

ilia-cher commented:
on each thread we have a sequence of coin flips, and coins might be different.
assume all coins have low probability (as of now)
after that we replace each coin with two coins, first flip a kLowProb coin and then flip the actual coin (if the first one is a success), the second flip is done with prob. sampling_prob_ / kLowProb, so overall prob. is sampling_prob_

now the sequence of first coin flips is the sequence of identical coins with prob. kLowProb, so we can replace it with sampling from geometric distribution (the probability distribution of the number X of Bernoulli trials needed to get one success)

Commit:
Update on "[rfc] Reduce number of coin flips in RecordFunction"


Summary:
Currently we flip a coin for each sampled callback each time
we run RecordFunction, this PR is an attempt to skip most of the coin
flips (for the low-probability observers) and keep the distribution
close to the original one

Test Plan:
CI and record_function_benchmark

[ghstack-poisoned]

Commit:
Update on "[rfc] Reduce number of coin flips in RecordFunction"


Summary:
Currently we flip a coin for each sampled callback each time
we run RecordFunction, this PR is an attempt to skip most of the coin
flips (for the low-probability observers) and keep the distribution
close to the original one

Test Plan:
CI and record_function_benchmark

[ghstack-poisoned]

ilia-cher commented:
updated benchmark and added benchmark results, as expected there's a reduction in time for the pure RecordFunction runtime

ngimel commented on binaries/record_function_benchmark.cc:
```diff
@@ -76,5 +76,30 @@ int main(int argc, char** argv) {
+      std::chrono::duration_cast<us>(clock::now() - start_time).count());
+  std::cout << "Pure RecordFunction runtime of " << FLAGS_rec_fn_iter
+            << " iterations " << duration
+            << " us, number of callback invocations: " << cb_count << std::endl;
```
maybe add expected number of callback invocations here?

Commit:
Update on "[rfc] Reduce number of coin flips in RecordFunction"


Summary:
Currently we flip a coin for each sampled callback each time
we run RecordFunction, this PR is an attempt to skip most of the coin
flips (for the low-probability observers) and keep the distribution
close to the original one

Test Plan:
CI and record_function_benchmark
```
(python_venv) iliacher@devgpu151:~/local/pytorch  (reduce_coin_flops)$ ./build/bin/record_function_benchmark
Warmup time: 30108 us.
Time per iteration (1x1): 1496.78 us.
Time per iteration (16x16): 2142.46 us.
Pure RecordFunction runtime of 10000000 iterations 687929 us, number of callback invocations: 978
(python_venv) iliacher@devgpu151:~/local/pytorch  (reduce_coin_flops)$ ./build/bin/record_function_benchmark
Warmup time: 19051 us.
Time per iteration (1x1): 1581.89 us.
Time per iteration (16x16): 2195.67 us.
Pure RecordFunction runtime of 10000000 iterations 682402 us, number of callback invocations: 1023
(python_venv) iliacher@devgpu151:~/local/pytorch  (reduce_coin_flops)$ ./build/bin/record_function_benchmark
Warmup time: 18715 us.
Time per iteration (1x1): 1566.11 us.
Time per iteration (16x16): 2131.17 us.
Pure RecordFunction runtime of 10000000 iterations 693571 us, number of callback invocations: 963
(python_venv) iliacher@devgpu151:~/local/pytorch  (reduce_coin_flops)$

(python_venv) iliacher@devgpu151:~/local/pytorch  (reduce_coin_flops)$ ./build/bin/record_function_benchmark
Warmup time: 18814 us.
Time per iteration (1x1): 1536.2 us.
Time per iteration (16x16): 1985.82 us.
Pure RecordFunction runtime of 10000000 iterations 944959 us, number of callback invocations: 1015
(python_venv) iliacher@devgpu151:~/local/pytorch  (reduce_coin_flops)$ ./build/bin/record_function_benchmark
Warmup time: 18278 us.
Time per iteration (1x1): 1526.32 us.
Time per iteration (16x16): 2093.77 us.
Pure RecordFunction runtime of 10000000 iterations 985307 us, number of callback invocations: 1013
(python_venv) iliacher@devgpu151:~/local/pytorch  (reduce_coin_flops)$ ./build/bin/record_function_benchmark
Warmup time: 18545 us.
Time per iteration (1x1): 1524.65 us.
Time per iteration (16x16): 2080 us.
Pure RecordFunction runtime of 10000000 iterations 952835 us, number of callback invocations: 1048
```


[ghstack-poisoned]

dzhulgakov reviewed (approved):
Just as a sanity check - can you run much more iterations and look at # of callback iterations?

From the sample size of 3 - it seems that all of them are larger than 1000, I'm just curious whether there's small bias there or it's just noise

ilia-cher commented:
it's not only >1000, added more runs in the test section

ilia-cher commented:
oh it was all 1000+ for the case when I rolled back the changes, for the PR version it was around 1000

ilia-cher commented:
hm, before this PR, it seems to be consistently more than 1000, maybe there was some bias, but now it seems to be on the both sides of 1000

Commit:
Update on "[rfc] Reduce number of coin flips in RecordFunction"


Summary:
Currently we flip a coin for each sampled callback each time
we run RecordFunction, this PR is an attempt to skip most of the coin
flips (for the low-probability observers) and keep the distribution
close to the original one

Test Plan:
CI and record_function_benchmark
```
(python_venv) iliacher@devgpu151:~/local/pytorch  (reduce_coin_flops)$ ./build/bin/record_function_benchmark
Warmup time: 30108 us.
Time per iteration (1x1): 1496.78 us.
Time per iteration (16x16): 2142.46 us.
Pure RecordFunction runtime of 10000000 iterations 687929 us, number of callback invocations: 978
(python_venv) iliacher@devgpu151:~/local/pytorch  (reduce_coin_flops)$ ./build/bin/record_function_benchmark
Warmup time: 19051 us.
Time per iteration (1x1): 1581.89 us.
Time per iteration (16x16): 2195.67 us.
Pure RecordFunction runtime of 10000000 iterations 682402 us, number of callback invocations: 1023
(python_venv) iliacher@devgpu151:~/local/pytorch  (reduce_coin_flops)$ ./build/bin/record_function_benchmark
Warmup time: 18715 us.
Time per iteration (1x1): 1566.11 us.
Time per iteration (16x16): 2131.17 us.
Pure RecordFunction runtime of 10000000 iterations 693571 us, number of callback invocations: 963
(python_venv) iliacher@devgpu151:~/local/pytorch  (reduce_coin_flops)$
(python_venv) iliacher@devgpu151:~/local/pytorch  (reduce_coin_flops)$ ./build/bin/record_function_benchmark
Warmup time: 21580 us.
Time per iteration (1x1): 1912.39 us.
Time per iteration (16x16): 2049.49 us.
Pure RecordFunction runtime of 10000000 iterations 651399 us, number of callback invocations: 999, expected number: ~1000 invocations
(python_venv) iliacher@devgpu151:~/local/pytorch  (reduce_coin_flops)$ ./build/bin/record_function_benchmark
Warmup time: 18947 us.
Time per iteration (1x1): 1591.05 us.
Time per iteration (16x16): 2199.92 us.
Pure RecordFunction runtime of 10000000 iterations 665978 us, number of callback invocations: 1014, expected number: ~1000 invocations
(python_venv) iliacher@devgpu151:~/local/pytorch  (reduce_coin_flops)$ ./build/bin/record_function_benchmark
Warmup time: 17512 us.
Time per iteration (1x1): 1455.78 us.
Time per iteration (16x16): 2055.89 us.
Pure RecordFunction runtime of 10000000 iterations 676929 us, number of callback invocations: 988, expected number: ~1000 invocations

before the PR coin flipping change:

(python_venv) iliacher@devgpu151:~/local/pytorch  (reduce_coin_flops)$ ./build/bin/record_function_benchmark
Warmup time: 18814 us.
Time per iteration (1x1): 1536.2 us.
Time per iteration (16x16): 1985.82 us.
Pure RecordFunction runtime of 10000000 iterations 944959 us, number of callback invocations: 1015
(python_venv) iliacher@devgpu151:~/local/pytorch  (reduce_coin_flops)$ ./build/bin/record_function_benchmark
Warmup time: 18278 us.
Time per iteration (1x1): 1526.32 us.
Time per iteration (16x16): 2093.77 us.
Pure RecordFunction runtime of 10000000 iterations 985307 us, number of callback invocations: 1013
(python_venv) iliacher@devgpu151:~/local/pytorch  (reduce_coin_flops)$ ./build/bin/record_function_benchmark
Warmup time: 18545 us.
Time per iteration (1x1): 1524.65 us.
Time per iteration (16x16): 2080 us.
Pure RecordFunction runtime of 10000000 iterations 952835 us, number of callback invocations: 1048
```

Differential Revision: [D22320879](https://our.internmc.facebook.com/intern/diff/D22320879)

[ghstack-poisoned]

ngimel commented on aten/src/ATen/record_function.cpp:
```diff
@@ -212,16 +212,58 @@ inline CallbackManager& manager() {
-double RecordFunctionCallback::sample_zero_one() {
+int sample_geometric() {
+  static thread_local auto gen =
+      std::make_unique<std::mt19937>(std::random_device()());
```
if random generation is still an issue, we can experiment with minstd_rand0 or knuth_b engines - we don't need high quality randoms here.
